src/librustc_typeck/check/wfcheck.rs

```diff
@@ -354,19 +343,90 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                 }
             }
 
-            let predicates = fcx.tcx.predicates_of(item_def_id).instantiate_identity(fcx.tcx);
-            let predicates = fcx.normalize_associated_types_in(item.span, &predicates);
-            this.check_where_clauses(fcx, item.span, &predicates);
+            this.check_where_clauses(fcx, item.span, item_def_id);
 
             fcx.impl_implied_bounds(item_def_id, item.span)
         });
     }
 
+    /// Checks where clauses and inline bounds that are declared on def_id.
     fn check_where_clauses<'fcx, 'tcx>(&mut self,
                                        fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
                                        span: Span,
-                                       predicates: &ty::InstantiatedPredicates<'tcx>)
-    {
+                                       def_id: DefId) {
+        use ty::subst::Subst;
+        use rustc::ty::TypeFoldable;
+
+        let mut predicates = fcx.tcx.predicates_of(def_id);
+        let mut substituted_predicates = Vec::new();
+
+        let generics = self.tcx.generics_of(def_id);
+        let is_our_default = |def: &ty::TypeParameterDef|
+                                def.has_default && def.index >= generics.parent_count() as u32;
+        let defaulted_params = generics.types.iter().cloned().filter(&is_our_default);
+        // Check that defaults are well-formed. See test `type-check-defaults.rs`.
+        // For example this forbids the declaration:
+        // struct Foo<T = Vec<[u32]>> { .. }
+        // Here the default `Vec<[u32]>` is not WF because `[u32]: Sized` does not hold.
+        for d in defaulted_params.map(|p| p.def_id) {
+            fcx.register_wf_obligation(fcx.tcx.type_of(d), fcx.tcx.def_span(d), self.code.clone());
```

Comment:
I'm confused -- won't this cause errors in cases like

```rust
struct Foo<T: Copy>() {
  t: T
}

struct Bar<T, U = Foo<T>> {
  t: T,
  u: U
}
```

Reply:
Yes good catch. I want this check because I feel that `Foo<T, U = <T as Iterator>::Item>` is a particularly weird way for a default to be conditionally applied, and might get (ab)used if we don't prevent it, but probably it's not really different from other cases involving dependent defaults.

For now I should do what you suggested above and do this check only if the type does not reference other type params.